### ISSUE TITLE
End to End Test should check Kubecontext before Proceeding

### ISSUE THREAD

**Issue by DerekTBrown** [kubernetes-sigs/kubebuilder#3971]
### What do you want to happen?

- Kubebuilder auto-generates tests which install/delete Prometheus Operator and Cert Manager Resources[^1]. Running these tests against a production cluster could lead to bad outcomes.
- These tests aren't labeled/put behind an interface that reflects their danger. For instance, running `make test` will execute these tests.
- It would be a good idea to guard against accidental invocation of these tests by checking the context before running them. For example, Tilt has some simplistic logic for checking that the target cluster is Kind/Minikube/Docker-desktop, and requiring users to explicitly provide input before proceeding against an unknown target cluster type. [^2]

[^1]: https://github.com/kubernetes-sigs/kubebuilder/blob/351d9e804bd348eb6a5b071493bc7e99b9b54050/pkg/plugins/golang/v4/scaffolds/internal/templates/test/e2e/test.go#L73
[^2]: https://docs.tilt.dev/api.html#api.allow_k8s_contexts

**Comment by camilamacedo86** [comment]
Hi @DerekTBrown,

Thank you for raise this one. Following the commands inline

> Kubebuilder auto-generates tests which install/delete Prometheus Operator and Cert Manager Resources[1](https://github.com/kubernetes-sigs/kubebuilder/issues/3971#user-content-fn-1-9e4dbba5563e3087283eb0864309283c). Running these tests against a production cluster could lead to bad outcomes

The propose of those tests are NOT to be executed against Production at all.
Those are e2e integration tests.
You should create a new env from the scratch to use those (as for any e2e tests) and keep it isolate
For example, using kind and GitHub actions.

You can see how we do in Kubebuilder the GitHub action (you would do quite the same) to ensure that those work
Example: https://github.com/kubernetes-sigs/kubebuilder/blob/master/.github/workflows/test-sample-go.yml

> These tests aren't labeled/put behind an interface that reflects their danger. For instance, running make test will execute these tests.

`make test` will **NOT** execute the e2e tests. See:

https://github.com/kubernetes-sigs/kubebuilder/blob/351d9e804bd348eb6a5b071493bc7e99b9b54050/testdata/project-v4/Makefile#L62-L64

The target which will execute those tests is: `make e2e-tests`

https://github.com/kubernetes-sigs/kubebuilder/blob/351d9e804bd348eb6a5b071493bc7e99b9b54050/testdata/project-v4/Makefile#L66-L69

Again, they are not dangerous. 
Those are only e2e tests.


> It would be a good idea to guard against accidental invocation of these tests by checking the context before running them. For example, Tilt has some simplistic logic for checking that the target cluster is Kind/Minikube/Docker-desktop, and requiring users to explicitly provide input before proceeding against an unknown target cluster type. [2](https://github.com/kubernetes-sigs/kubebuilder/issues/3971#user-content-fn-2-9e4dbba5563e3087283eb0864309283c)

The tests cannot be called by your users (Operator/Manager consumers)
The tests can be ONLY called you who is maintainer of your project and when you see that fit.
It is not possible the tool decide what context of Kubenetes you should or not use to run the tests. 

I hope that the info above clarifies.
What seems the issue here is that someone change the default implementation scaffold by Kubebuilder and is not using appropriated. That is not something that we can sort out. 

So, I am closing this one.
However, if you think that it should be re-opened please feel free to do so.

**Comment by majewsky** [comment]
> The propose of those tests are NOT to be executed against Production at all.

Please consider that people make mistakes.

In our QA cluster, we are currently restoring our Prometheus and cert-manager resources from backup for the second time in a month.

**Comment by camilamacedo86** [comment]
Hi @majewsky 

> Please consider that people make mistakes.

I am not sure how we would make clear that e2e tests should not be executed against production envs
Run e2e tests against production envs does not seems a best practice and a good idea in any scenario. 

However, I think we might have a hall for improvements here.
- Create a GitHub action that will call the e2e tests as we do in: https://github.com/kubernetes-sigs/kubebuilder/blob/master/.github/workflows/test-sample-go.yml 
- The above action should be scaffold in the project. It should make clear how to use it
- Add comments in the e2e tests scaffolded, and in the GitHub action scaffolded that try clarifies its intentions and purposes. 

I am re-open this one to see if we can do the suggested above improvements
Please, feel free to contribute with it if you wish
Your collaboration is very welcome.

**Comment by majewsky** [comment]
> Run e2e tests against production envs does not seems a best practice and a good idea in any scenario. We have no place where we recommend, suggest or encourage this approach.

Yes, but that's besides the point. If there is `make test` in some repository, then inevitably at least a certain amount of people are going to assume that it is safe to run without reading documentation in another repository first. This would not be the case if it were called something like `make test-and-then-blow-up-the-cluster` (which I'm not saying as a suggestion, just to make a point).

The suggestions you provide about scaffolding a GitHub action are ineffective insofar as they don't change the existence of `make test` that can blow up whatever your kubeconfig happens to point to.

OP already provided some suggestions how to better address this:

1. "These tests aren't labeled/put behind an interface that reflects their danger." That's what my tongue-in-cheek example of `make test-and-then-blow-up-the-cluster` was refering to. An actual implementation might look more like the test printing at the end something like `Hey, I did not delete the CRDs that I added. Please execute the test with CLEANUP_PROMETHEUS_CRDS=1 and CLEANUP_CERTMANAGER_CRDS=1 to do so.` That would make the impact clear.
2. "checking the context before running them." As explained, if it's not a variant of Kubernetes deployment that is commonly used for unit tests, that should be grounds for suspicion and require an extra confirmation to run, i.e. there would be a message like `This test contains destructive steps and should run inside of envtest [or whatever], but it looks like this is a regular Kubernetes cluster. If you want to run the test here and accept the risk, execute the test with I_KNOW_THIS_WILL_BLOW_UP_STUFF=1.`

I'm going to add my own suggestion:

3. A heuristic approach would be to refuse to delete the CRDs if there are still resources associated with them in namespaces that are unrelated to the test procedure(s).

And finally, one thing that is more of an observation than a readymade suggestion. I am coming at this from the perspective of a cluster admin, not as an operator developer, so I don't know what's going on inside of kubebuilder. But as far as I understand, when a developer uses kubebuilder to scaffold an operator, it puts something in `make test` that messes with the Prometheus and cert-manager CRDs. Why is that needed at all? If it's about providing a template that developers can replace with their own CRDs, could this be replaced with an `example.com`-type dummy CRD that will not break if it accidentally gets deployed into (and then undeployed from) an existing cluster?

**Comment by camilamacedo86** [comment]
Hi @majewsky,

I believe there are a few misunderstandings:

- a) make test does not run e2e tests.

https://github.com/kubernetes-sigs/kubebuilder/blob/94d8fce49227f87f7a4e79f9c5d121e1e942b5bf/testdata/project-v4/Makefile#L62-L64

- b) The e2e tests or unit tests do not blow up the cluster at all. However, it's common sense to run e2e tests against an isolated environment created from scratch, which will be destroyed afterward.

- c) The e2e tests have their own target.

https://github.com/kubernetes-sigs/kubebuilder/blob/94d8fce49227f87f7a4e79f9c5d121e1e942b5bf/testdata/project-v4/Makefile#L66-L69

- d) If you run `make help`, you'll see that all test targets are **under the development section**

- e) The project scaffold is meant for development environments and should not be distributed to end users. So, end users would not have the make targets commands at all.  Please refer to the README scaffold in your project and search for solutions like OperatorHub.io. 

While there's always room for improvement, there are limits to what we can do if users run commands against production environments without first understanding what they do or reading the documentation.

_In other words, if you've configured your dev environment for production and then run make test, you should at least know what the tests are doing. Unfortunately, this is an issue that we can't completely solve. We do our best to upskill people, but some level of common sense and knowledge is expected from developers when they are adopting any framework/tool._

**Comment by camilamacedo86** [comment]
Hi @majewsky,

Thank you a lot for your input. But, I believe there are a few misunderstandings:

- a) make test does not run e2e tests.

https://github.com/kubernetes-sigs/kubebuilder/blob/94d8fce49227f87f7a4e79f9c5d121e1e942b5bf/testdata/project-v4/Makefile#L62-L64

- b) The e2e tests or unit tests do not blow up the cluster at all. However, it's common sense to run e2e tests against an isolated environment created from scratch, which will be destroyed afterward.

![Screenshot 2024-08-21 at 14 38 40](https://github.com/user-attachments/assets/fddbe495-7ab3-45fa-aca4-27dbc4872ce7)

Ditto so, the e2e-test will install what is required to be executed. However, if the dev uses it against a pre-exist env then, then can customize their tests accordingly. 

- c) The e2e tests have their own target.

https://github.com/kubernetes-sigs/kubebuilder/blob/94d8fce49227f87f7a4e79f9c5d121e1e942b5bf/testdata/project-v4/Makefile#L66-L69

- d) If you run `make help`, you'll see that all test targets are **under the development section**

- e) The project scaffold is meant for development environments and should not be distributed to end users. So, end users would not have the make targets commands at all.  Please refer to the [README](https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v4/README.md#project-distribution) scaffold in your project and search for solutions like [OperatorHub.io](https://operatorhub.io/). You can either look for all a lot of examples of other projects which uses Kubebuilder. Nobody distributes the source code of the project. That is valid only for maintainers.


By last, regards your request:

> The tests cannot be called by your users (Operator/Manager consumers)
> The tests can be ONLY called you who is maintainer of your project and when you see that fit.
> It is not possible the tool decide what context of Kubenetes you should or not use to run the tests.

**We cannot address them because they are common sense**

Nobody expected that end users would run tests, and nobody expected that any target would automatically decide the kubectl context. It can not find this addressed in any solution available. 

> _While there's always room for improvement, there are limits to what we can do if users run commands against production environments without first understanding what they do or reading the documentation. We do our best to help upskill, but some level of common sense and knowledge is expected from developers when they are adopting any framework/tool._

**But has one thing that we might can do here:**

Add a logic to not install the Prometheus and CertManager if when for example an envvar be set as false, i.e. `INSTALL_PROMETHEUS_E2E_TEST=false`. But users will still need to checking what means call the targets and where/when they should call it safe for their specific envs.

**Comment by camilamacedo86** [comment]
In line with the takeaways from this discussion, we’re moving forward with the following:

- Enhance the e2e test scaffolding and add a GitHub Action to demonstrate their usage (#4088)
- Enhance E2E Test Scaffold with Optional Prometheus and Cert-Manager Installation (#4089)

Beyond these improvements, I'm finding it challenging to identify additional areas for enhancement. However, if you have any ideas or suggestions, please feel free to raise a new issue for each specific suggestion or idea.

**Comment by camilamacedo86** [comment]
**Just to Supplement I hope that can contribute and help out**
**Also to bring some conclusion**

Hi @majewsky,

> I’d like to offer some additional input to help address the situation. It seems you manage a shared environment for your developers, and it appears that someone mistakenly executed `make test-e2e` in this environment, resulting in deletions or inaccurate changes that affected the entire environment for all developers.

### How People Usually Solve This Issue:

- **(Role Devs) User Customization:** Developers should customize the default scaffold for their specific needs. This is why we include `TODO(user)` comments. In this case, improving the comments and documentation is something we’re addressing in [#4088](https://github.com/kubernetes-sigs/kubebuilder/issues/4088).
- **(Role System Admin): RBAC Implementation:** Admins typically use Role-Based Access Control (RBAC) to define roles and permissions, restricting changes in shared environments, especially in specific namespaces or non-isolated environments. Ensure that development and testing activities, like running `make test-e2e`, are performed in isolated environments. Lock down production or shared environments to prevent unintended modifications.
- **(Role System Admin): Dedicated Namespaces:** Use separate namespaces for different teams or projects. Critical resources like Prometheus can be deployed in a protected namespace that is not accessible to developers.
- **(End users or consumers of the solution) Running `make test-e2e`:** End users or consumers of the solution should not have access to the source code. As a result, they should not be able to execute any `make` commands. The source code is for maintainers, and there are many ways to distribute the solution. Kubebuilder provides a helper for that. See: [Project Distribution](https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v4/README.md#project-distribution). Many people also use Helm Charts for this purpose. Due that we have in our roadmap delivery a plugin to help on it, see: https://github.com/kubernetes-sigs/kubebuilder/pull/3632


Hi @DerekTBrown,

Please see the points above. Additionally, please note that any target that deploys, undeploys, or updates resources in the cluster can potentially cause issues if executed against the wrong environment. Therefore, we cannot change the standard terms like `make test-e2e` or `make test`, which are intuitive and descriptive, to something like `make test-and-then-blow-up-the-cluster`.

**Indeed, e2e tests do not inherently "blow up" the cluster;** this depends on the environment they are run against. It’s crucial to follow best practices, such as running e2e tests **only** against isolated environments, which are often set up specifically for this purpose _(often you spinning up, create the env run the test and then teardown - i.e. in CI checks to validate changes in the PR which is the intention of)_.

Additionally, we cannot determine the safety of specific vendors or environments, as these are unique to each setup. However, we will work on improving the comments and documentation around these processes. We’ve also created a task to explore how we can enhance the cleanup processes in e2e tests and make them more transparent. See: https://github.com/kubernetes-sigs/kubebuilder/issues/3971#issuecomment-2302422739

I hope that clarifies and can help out. 
Thank you a lot for your contribution and inputs.

Please feel free to open new issues if you need to do so.

**Comment by camilamacedo86** [comment]
It was either discussed in the channel: https://kubernetes.slack.com/archives/CAR30FCJZ/p1724837864980059

**Comment by camilamacedo86** [comment]
Hi @DerekTBrown and @majewsky,

I just wanted to let you know that this scenario is addressed in master and will be available in the next release. Here's how it has been addressed:

- **a)** The tests are done using Kind, so in the Makefile we are checking if Kind is up and running:  

https://github.com/kubernetes-sigs/kubebuilder/blob/fa9f70ce2f59b56ec69fccd4e708afef09ca94d3/testdata/project-v4/Makefile#L66-L81

- **b)** Also, we are loading the image before installing the dependencies, so if the tests are running in an environment that is not Kind, it will fail:  

https://github.com/kubernetes-sigs/kubebuilder/blob/fa9f70ce2f59b56ec69fccd4e708afef09ca94d3/testdata/project-v4/test/e2e/e2e_suite_test.go#L78-L82

- **c)** We are checking in the e2e tests if the CRDs are found on the cluster to skip the install and avoid the scenario:  

https://github.com/kubernetes-sigs/kubebuilder/blob/fa9f70ce2f59b56ec69fccd4e708afef09ca94d3/testdata/project-v4/test/e2e/e2e_suite_test.go#L84-L107

Lastly, a lot of improvements have been made in this area to provide more value to the users. The e2e tests now offer a very comprehensive set of checks by default, and for common basic scenarios, users mainly need to apply the CR and check its results via metrics or status. Those enhancements were made since we ware looking to see how to improve it based in this issue and scenarios that you both reported.

I hope that helps and please feel free to raise issues / suggestions for it or anything else that you might see fit.

Again, thank you for your contribution 🥇
